### PR TITLE
feat: support edge runtimes

### LINKED ISSUE #38: feat: support Vercel Edge Runtime
Hey Knock team! Knock's been featured in a few places by Vercel, and my team is leaning heavily on the Vercel Edge Runtime with Next.js, which doesn't support the full Node.js ecosystem. The `@knocklabs/node` package uses `axiom` and `jsonwebtoken` which are not compatible. Can you guys please build in support for the Edge Runtime by switching to the Web Fetch API and [jose](https://www.npmjs.com/package/jose)? Might require some fallbacks for Node.js runtimes before v17, which don't include `fetch`.

### PR DESCRIPTION
**Pull request by connorlindsey**
This PR makes the Node SDK compatible with edge runtimes including Vercel Edge functions. The main changes include:
1. Replace Axios with a native fetch wrapper
2. Replace jsonwebtoken with jose as jsonwebtoken uses Node APIs like Crypto which aren't supported by the edge runtime

Note: This could use a fair amount of testing, especially around error handling. Here's an repo with a test Next 13 app which has an edge function to test against: https://github.com/knocklabs/next-edge-test. You can also use `npx ts-node test.ts` to run a typescript script and run the client directly.

Post-release: Bump minor version and announce breaking change in release notes

Resolves #38

### CODE REVIEW

File: src/common/fetchClient.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+export interface FetchClientConfig {
+  baseURL: string;
+  headers: Record<string, string>;
```
Comment (thomaswhyyou) [inline src/common/fetchClient.ts:3]: Small thing/feel free to ignore but wondering if headers must be required for the client config, though we do pass one in where we use it.

File: src/knock.ts [hunk 2]
```diff
@@ -77,7 +77,7 @@ class Knock {
    * @param options Optionally specify the signing key to use (in PEM or base-64 encoded format), and how long the token should be valid for in seconds
    * @returns {string} A JWT token that can be used to authenticate requests to the Knock API (e.g. by passing into the <KnockFeedProvider /> component)
    */
-  static signUserToken(userId: string, options: SignUserTokenOptions) {
+  static async signUserToken(userId: string, options: SignUserTokenOptions) {
```
Comment (JEverhart383) [inline src/knock.ts:80]: Moving `signUserToken` from a sync to async function is the only breaking change I can see here. Should we update the JSDoc return type to be a promise to make this more clear? `@returns {Promise<string>}` or maybe consider using the `@async` tag: https://jsdoc.app/tags-async.

File: src/knock.ts [hunk 3]
```diff
@@ -86,28 +86,29 @@ class Knock {
...
-        algorithm: "RS256",
-      },
-    );
+    return await new SignJWT({
+      sub: userId,
+      iat: currentTime,
+      exp: currentTime + expireInSeconds,
+    })
+      .setProtectedHeader({ alg: "RS256", typ: "JWT" })
+      .sign(new TextEncoder().encode(signingKey));
```
Comment (JEverhart383) [inline src/knock.ts:95]: I pulled down the example Next app to play around with, and get this error when trying to generate a signing key. Entirely possible I'm doing something wrong though: Screenshot 2024-01-05 at 1 14 49 PM

**Review by JEverhart383 — COMMENTED** [review]
Left a few comments. Did the example app actually get tested in the edge runtime? I saw the deploy to Vercel failed (I suspect b/c of the relative file import). It might also be worth testing this on Cloudflare, if Vercel's edge doesn't use theme behind the scenes.

File: src/common/fetchClient.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  ): Promise<FetchResponse> {
+    const url = this.buildUrl(path, config.params);
+
+    const response = await fetch(url, {
+      method,
+      headers: {
+        ...this.config.headers,
+        ...(config.headers ?? {}),
+      },
+      body: config.body ? JSON.stringify(config.body) : undefined,
```
Comment (thomaswhyyou) [inline src/common/fetchClient.ts:36]: I know it probably doesn't really make a difference in our usage of this fetch clicent, but is it right to always assume to JSON.stringfy the body data? Wouldn't it be dependent on the http method and also the type of data we are sending?

File: src/common/fetchClient.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    const contentType = response.headers.get("content-type");
+    if (contentType && contentType.includes("application/json")) {
+      data = await response.json();
+    } else if (contentType && contentType.includes("text")) {
+      data = await response.text();
+    } else {
+      data = await response.blob();
+    }
+
+    return data;
```
Comment (thomaswhyyou) [inline src/common/fetchClient.ts:93]: More of a question, but is it be bad to fall back to blob() when `contentType` is not present, instead of returning `undefined`?

File: src/knock.ts [hunk 6]
```diff
@@ -153,7 +155,7 @@ class Knock {
   }
 
   handleErrorResponse(path: string, error: any) {
-    if (axios.isAxiosError(error) && error.response) {
+    if (error.response) {
```
Comment (thomaswhyyou) [inline src/knock.ts:158]: Just want to double check - will this handler still get called for non-2xx responses? Asking because I'm not sure if fetch automatically handles non-2xx responses by throwing therefore catching as an error in this.get/post/put/delete methods. It [sounds like](https://web.dev/articles/fetch-api-error-handling#when_the_network_status_code_represents_an_error) that may not be the case, hence wanting to double check.

)

**Review by thomaswhyyou — COMMENTED** [review]
Left a couple comments/questions, overall this makes sense to me.

File: src/knock.ts (hunk 2 above)
Comment (connorlindsey) [reply src/knock.ts:80]: Good call out, I'll update the return type

File: src/common/fetchClient.ts (hunk 1 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts:3]: Yeah, technically neither are required so I'll make them optional.

File: src/common/fetchClient.ts (hunk 4 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts:36]: Looked at the Axios source and you're right. They do some more checks around the header type and checking if the body is an object, form data, etc. I'll look at adding a basic check for the body.

File: src/common/fetchClient.ts (hunk 5 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts:93]: This was one thing I wasn't sure on. I think some sort of fallback is a good idea, maybe we just fall back to sending the raw response text if we don't have a good idea of the correct type.

File: src/knock.ts (hunk 6 above)
Comment (connorlindsey) [reply src/knock.ts:158]: Fetch doesn't, so if we're relying on that behavior from axios, we'll need to add that back in. I'll address that in my next commit.

File: src/common/fetchClient.ts (hunk 5 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts:93]: Looking into it more, I think blob is a decent fallback based on the available methods: https://developer.mozilla.org/en-US/docs/Web/API/Response#instance_methods

File: src/common/fetchClient.ts (hunk 4 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts]: I added a `prepareRequestBody` function which checks the headers and input data to format the request.

File: src/knock.ts (hunk 6 above)
Comment (connorlindsey) [reply src/knock.ts:158]: Updated to throw on non-2xx errors by checking `response.ok`. Tested it with a 401 error by providing an invalid key and the error flows through correctly still.

**Comment by thomaswhyyou** [comment]
Taking another look 👀

File: src/common/fetchClient.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+export interface FetchResponseErrorData<T = any> {
+  status: number;
+  headers: Record<string, string>;
+  data: T;
+}
+
+export class FetchResponseError extends Error {
+  readonly response: FetchResponseErrorData;
+
+  constructor(response: FetchResponseErrorData) {
```
Comment (thomaswhyyou) [inline src/common/fetchClient.ts:25]:
```suggestion
  constructor(response: FetchResponse) {
```
Continuing on from comment/question re: `this.headersToObject()` below - if we can skip and just use `headers.get()`, then I'm guessing we can use `FetchResponse` here without needing to define a new `FetchResponseErrorData` type?

File: src/common/fetchClient.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  private prepareRequestBody(
+    data: any,
+    headers: Record<string, string>,
+  ): string | FormData {
+    if (headers?.["Content-Type"]?.includes("application/json")) {
+      if (typeof data === "string") {
+        return data;
+      }
```
Comment (thomaswhyyou) [inline src/common/fetchClient.ts:118]: Can I double check what scenario this nested conditional here is intended to handle? Asking because I was getting a bit confused following the conditionals flow here. For example, if the content type was set to "text/plain", and the data passed in was a string, we would still JSON.stringify() and wasn't sure if that's the expected behavior. 

One other thing that jumped out was here we check the content type header specifically for "application/json", but not for "multipart/form-data" when handling `FormData` below?

I was wondering if we can simplify the logic here, and do the thing based on the data passed in? Something like:
```
if (!data) return undefined;
if (typeof data === "string") return data;
if (data instance of FormData) return data;
return JSON.stringify(data);
```

I may be missing some nuances here, so let me know what you think!

**Review by thomaswhyyou — COMMENTED** [review]
This is looking good, sorry the delay! I did leave a few follow up comments/questions. One main question among them is whether we can avoid a breaking change with the `signUserToken` method?

File: src/knock.ts (hunk 3 above)
Comment (thomaswhyyou) [reply src/knock.ts]: @JEverhart383 Jeff do you think this is something you could take another look/check to see if you are still getting the error above?

File: src/knock.ts [hunk 9]
```diff
@@ -75,9 +75,9 @@ class Knock {
    *
    * @param userId {string} The ID of the user that needs a token, e.g. the user viewing an in-app feed.
    * @param options Optionally specify the signing key to use (in PEM or base-64 encoded format), and how long the token should be valid for in seconds
-   * @returns {string} A JWT token that can be used to authenticate requests to the Knock API (e.g. by passing into the <KnockFeedProvider /> component)
+   * @returns {Promise<string>} A JWT token that can be used to authenticate requests to the Knock API (e.g. by passing into the <KnockFeedProvider /> component)
```
[the remark this answers is not in the document]
Comment (connorlindsey) [reply src/knock.ts:78]: I don't think it's possible to rewrite it to be synchronous because the underlying web crypto api's are asynchronous: https://github.com/panva/jose/discussions/532

File: src/common/fetchClient.ts (hunk 8 above)
Comment (connorlindsey) [reply src/common/fetchClient.ts:118]: Whoops, must've messed up the logic while I was refactoring it a bit. The goal was to check if the data was already stringified, but I think the logic you suggested is a lot simpler and understandable. I'll use that instead

File: src/common/fetchClient.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        ? this.prepareRequestBody(config.body, headers)
+        : undefined,
+    });
+    const data = await this.getResponseData(response);
+
+    if (!response.ok) {
+      throw new FetchResponseError({
+        status: response.status,
+        data,
+        headers: this.headersToObject(response.headers),
```
[the remark this answers is not in the document]
Comment (connorlindsey) [reply src/common/fetchClient.ts:70]: Yeah, it was only for compatibility for when we're checking the error / request-id header. Good call out, I'll just update that function to accept actual headers which'll simplify the code a bit.

File: src/knock.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (thomaswhyyou) [reply src/knock.ts:78]: Got it okay hm that's unfortunate - we should probably flag/socialize this to the team before cutting a new release.

File: src/knock.ts (hunk 3 above)
Comment (connorlindsey) [reply src/knock.ts]: Tested and this is fixed

**Review by thomaswhyyou — APPROVED** [review]
This looks good to go! 🙏

**Comment by heymartinadams** [comment]
Getting this error in production for `v0.6.0`, and am unable to debug it because the error message isn’t clear. No errors for `0.5.1`.

image

Code:
image

**Comment by connorlindsey** [comment]
Hey @heymartinadams, thanks for reporting this issue, we'll look into it this week. 

For ease of tracking it, could you open this as a separate issue in this repo? Also, it'd be big help if you could include that code snippet as text instead of an image.
